### PR TITLE
Add byte quantization for float vectors in HNSW

### PR DESCRIPTION
**Pull request by benwtrent**
Adds new `quantization_options` to `dense_vector`. This allows for vectors to be automatically quantized to `byte` when indexed.

Example:
```
PUT vectors
{
  "mappings": {
    "properties": {
      "my_vector": {
        "type": "dense_vector",
        "index": true,
        "index_options": {
          "type": "int8_hnsw"
        }
      }
    }
  }
}
```

When querying, the query vector is automatically quantized and used when querying the HNSW graph. This reduces the memory required to only `25%` of what was previously required for `float` vectors at a slight loss of accuracy.

This is currently only available when `index: true` and when using `hnsw`

### CODE REVIEW

File: docs/reference/how-to/knn-search.asciidoc [hunk 1]
```diff
@@ -66,6 +71,14 @@ reduction techniques like PCA. When experimenting with different approaches,
 it's important to measure the impact on relevance to ensure the search
 quality is still acceptable.
 
+[discrete]
+=== Reduce vector memory foot-print
+
+As of 8.12 the default <<dense-vector-element-type,`element_type`>> is `float`. But this can be
+automatically quantized during index time through the <<dense-vector-quantization,`quantization`>>. Quantization will
+reduce the required memory by 4x, but it will also reduce the precision of the vectors. For `float` vectors with
```
Comment (kderusso) [inline docs/reference/how-to/knn-search.asciidoc:79]: Can we link to any information (blog maybe?) here that can explain the space/precision tradeoffs?

File: docs/reference/mapping/types/dense-vector.asciidoc [hunk 2]
```diff
@@ -111,6 +111,39 @@ PUT my-index-2
 efficient kNN search. Like most kNN algorithms, HNSW is an approximate method
 that sacrifices result accuracy for improved speed.
 
+[[dense-vector-quantization]]
+==== Automatically quantize vectors for kNN search
+
+The `dense_vector` type supports quantization to reduce the memory footprint required when searching `float` vectors.
+Currently the only quantization method supported is `byte` and provided vectors `element_type` must be `float`.
```
Comment (kderusso) [inline docs/reference/mapping/types/dense-vector.asciidoc:118]: I think we want to add a line here that this is only supported for HNSW right?

File: docs/reference/mapping/types/dense-vector.asciidoc [hunk 3]
```diff
@@ -212,14 +245,34 @@ properties must be defined.
...
-(Required, integer)
+(Optional, integer)
 The number of candidates to track while assembling the list of nearest
 neighbors for each new node. Defaults to `100`.
+
+`quantization_options`:::
+(Optional, object)
+An optional section that configures the quantization configuration. The
+quantization configuration is used to reduce the memory footprint of the
+index. Only `byte` quantization is currently supported and can only be configured if
```
Comment (kderusso) [inline docs/reference/mapping/types/dense-vector.asciidoc:261]: Same note about HNSW

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 4]
```diff
@@ -171,6 +172,17 @@ public Builder(String name, IndexVersion indexVersionCreated) {
                     }
                 }
             });
+            this.indexOptions.addValidator(v -> {
+                if ((v instanceof HnswIndexOptions hnswIndexOptions)
+                    && hnswIndexOptions.quantizationOptions != null
+                    && elementType.getValue() == ElementType.BYTE) {
+                    throw new IllegalArgumentException(
+                        "[element_type] cannot be [byte] when using quantization type ["
```
Comment (kderusso) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:180]: Nitpick: Error message should state that float is the only one that is allowed (Uno reverse)

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/41_knn_search_byte_quantized.yml [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+              name:
+                type: keyword
+              vector:
+                type: dense_vector
+                dims: 5
+                index: true
+                similarity: l2_norm
+                index_options:
+                  type: hnsw
+                  quantization_options:
```
Comment (kderusso) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/41_knn_search_byte_quantized.yml:20]: Do we want to add any tests validating errors if we try to create quantized indices with unsupported values?

**Review by kderusso — COMMENTED** [review]
This looks awesome!!

File: docs/reference/how-to/knn-search.asciidoc [hunk 6]
```diff
@@ -66,6 +71,14 @@ reduction techniques like PCA. When experimenting with different approaches,
 it's important to measure the impact on relevance to ensure the search
 quality is still acceptable.
 
+[discrete]
+=== Reduce vector memory foot-print
+
+As of 8.12 the default <<dense-vector-element-type,`element_type`>> is `float`. But this can be
```
Comment (jimczi) [inline docs/reference/how-to/knn-search.asciidoc:77]: The doc is per version so not sure if it's worth mentioning `8.12`?

File: docs/reference/search/search-your-data/knn-search.asciidoc [hunk 7]
```diff
@@ -242,6 +242,114 @@ POST byte-image-index/_search
...
+[[knn-search-quantized-example]]
+==== Byte quantized kNN search
+
+If you want to provide `float` vectors, but want the memory savings of `byte` vectors, you can use the
+<<dense-vector-quantization, quantization>> feature. Quantization allows you to provide `float` vectors, but
+internally they are indexed as `byte` vectors. Additionally, the original `float` vectors are still retained
+in the index. This allows you to use the original `float` vectors for re-scoring, but the `byte` vectors for
+indexing.
+
+To use quantization, you must provide a `quantization_params` object in the `dense_vector` mapping.
```
Comment (jimczi) [inline docs/reference/search/search-your-data/knn-search.asciidoc:255]: It should refer to `quantization_options`?

File: docs/reference/search/search-your-data/knn-search.asciidoc [hunk 8]
```diff
@@ -242,6 +242,114 @@ POST byte-image-index/_search
...
+PUT quantized-image-index
+{
+  "mappings": {
+    "properties": {
+      "image-vector": {
+        "type": "dense_vector",
+        "element_type": "float",
+        "dims": 2,
+        "index": true,
+        "quantization_params": {
```
Comment (jimczi) [inline docs/reference/search/search-your-data/knn-search.asciidoc:268]: same here?

File: docs/reference/mapping/types/dense-vector.asciidoc [hunk 9]
```diff
@@ -212,14 +245,34 @@ properties must be defined.
...
+=====
+
+`type`:::
+(Required, string)
+The type of quantization. Currently only `byte` is supported.
+
+`confidence_interval`:::
+(Optional, float)
+The confidence interval to use when quantizing the vectors, can be any value between and including `0.90` and `1.0`.
+Defaults to `1/(dims + 1)`.
```
Comment (jimczi) [inline docs/reference/mapping/types/dense-vector.asciidoc:274]: Maybe explain what the confidence interval does? And also why not using `configured_quantile` like in the format ctr?

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java [hunk 10]
```diff
@@ -702,29 +714,74 @@ private abstract static class IndexOptions implements ToXContent {
...
         private final int efConstruction;
+        private final Int8ScalarQuantizationOptions quantizationOptions;
 
         static IndexOptions parseIndexOptions(String fieldName, Map<String, ?> indexOptionsMap) {
             Object mNode = indexOptionsMap.remove("m");
             Object efConstructionNode = indexOptionsMap.remove("ef_construction");
+            Object quantizationOptionsNode = indexOptionsMap.remove("quantization_options");
             if (mNode == null) {
-                throw new MapperParsingException("[index_options] of type [hnsw] requires field [m] to be configured");
+                mNode = Lucene99HnswVectorsFormat.DEFAULT_MAX_CONN;
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:752]: We probably need a test that configures `m` and not `ef_construction` since it was not possible before this change?

**Review by jimczi — COMMENTED** [review]
That's great @benwtrent !
I left some minor comments.
We'll also need to update the `tune-knn-search` docs but that can be done in a follow up.

File: docs/reference/how-to/knn-search.asciidoc (hunk 6 above)
Comment (benwtrent) [reply docs/reference/how-to/knn-search.asciidoc:77]: DOH! for sure, docs are already versioned!

File: docs/reference/search/search-your-data/knn-search.asciidoc (hunk 7 above)
Comment (benwtrent) [reply docs/reference/search/search-your-data/knn-search.asciidoc:255]: yep!

File: docs/reference/mapping/types/dense-vector.asciidoc (hunk 9 above)
Comment (benwtrent) [reply docs/reference/mapping/types/dense-vector.asciidoc:274]: @jimczi I think we should rename it in the ctor. Its really the confidence interval. I will rename it in Lucene

Comment (benwtrent) [reply docs/reference/mapping/types/dense-vector.asciidoc:274]: Let me get @tveasey 's opinion. What do you think. `confidence_interval` or `quantile`?

Comment (tveasey) [reply docs/reference/mapping/types/dense-vector.asciidoc:274]: As we discuss off line:

The q'th quantile x is a single point in the distribution curve, i.e. a value that x such that F(x) = q. A (central) p'th confidence interval is a pair of values (x, y) such that F(x) = 1 - p/2 and F(y) = 1 + p/2

So I believe strictly this is working in terms of a confidence interval since we find a pair or values we use to clamp values.

File: docs/reference/mapping/types/dense-vector.asciidoc (hunk 2 above)
Comment (benwtrent) [reply docs/reference/mapping/types/dense-vector.asciidoc:118]: Maybe? But the fact that its under the `index_options` already implies this.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java (hunk 4 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:180]: Maybe? I am also thinking ahead for when we add half_float support, which will support int8 quantization.

**Comment by benwtrent** [comment]
Did some rally tests (this is without force-merging), this is the `so_vector` dataset, so its 2M 768 float32 vectors, so requiring about 6240000000 bytes or ~6GB of off heap if using float. But quantizing to byte, it only requires about 1.5GB.

You can see how script score over all the vectors hits many page faults as the data doesn't fit in memory.

Also note all numbers reflect going from California (where the test cluster was located) to east coast (where my rally machine is).

```
------------------------------------------------------
    _______             __   _____
   / ____(_)___  ____ _/ /  / ___/_________  ________
  / /_  / / __ \/ __ `/ /   \__ \/ ___/ __ \/ ___/ _ \
 / __/ / / / / / /_/ / /   ___/ / /__/ /_/ / /  /  __/
/_/   /_/_/ /_/\__,_/_/   /____/\___/\____/_/   \___/
------------------------------------------------------
|                                                         Metric |                                Task |         Value |   Unit |
|---------------------------------------------------------------:|------------------------------------:|--------------:|-------:|
|                                                 Min Throughput |          knn-search-10-50-match-all |  10.06        |  ops/s |
|                                                Mean Throughput |          knn-search-10-50-match-all |  10.63        |  ops/s |
|                                              Median Throughput |          knn-search-10-50-match-all |  10.67        |  ops/s |
|                                                 Max Throughput |          knn-search-10-50-match-all |  11.07        |  ops/s |
|                                        50th percentile latency |          knn-search-10-50-match-all |  78.8566      |     ms |
|                                        90th percentile latency |          knn-search-10-50-match-all |  81.0507      |     ms |
|                                        99th percentile latency |          knn-search-10-50-match-all |  87.6091      |     ms |
|                                       100th percentile latency |          knn-search-10-50-match-all |  87.7286      |     ms |
|                                   50th percentile service time |          knn-search-10-50-match-all |  78.8566      |     ms |
|                                   90th percentile service time |          knn-search-10-50-match-all |  81.0507      |     ms |
|                                   99th percentile service time |          knn-search-10-50-match-all |  87.6091      |     ms |
|                                  100th percentile service time |          knn-search-10-50-match-all |  87.7286      |     ms |
|                                                     error rate |          knn-search-10-50-match-all |   0           |      % |
|                                                 Min Throughput |        script-score-query-match-all |   1.66        |  ops/s |
|                                                Mean Throughput |        script-score-query-match-all |   1.68        |  ops/s |
|                                              Median Throughput |        script-score-query-match-all |   1.68        |  ops/s |
|                                                 Max Throughput |        script-score-query-match-all |   1.69        |  ops/s |
|                                        50th percentile latency |        script-score-query-match-all | 573.453       |     ms |
|                                        90th percentile latency |        script-score-query-match-all | 587.929       |     ms |
|                                        99th percentile latency |        script-score-query-match-all | 618.025       |     ms |
|                                       100th percentile latency |        script-score-query-match-all | 653.512       |     ms |
|                                   50th percentile service time |        script-score-query-match-all | 573.453       |     ms |
|                                   90th percentile service time |        script-score-query-match-all | 587.929       |     ms |
|                                   99th percentile service time |        script-score-query-match-all | 618.025       |     ms |
|                                  100th percentile service time |        script-score-query-match-all | 653.512       |     ms |
|                                                     error rate |        script-score-query-match-all |   0           |      % |
|                                                 Min Throughput |   knn-search-10-50-acceptedAnswerId |  10.27        |  ops/s |
|                                                Mean Throughput |   knn-search-10-50-acceptedAnswerId |  10.41        |  ops/s |
|                                              Median Throughput |   knn-search-10-50-acceptedAnswerId |  10.42        |  ops/s |
|                                                 Max Throughput |   knn-search-10-50-acceptedAnswerId |  10.52        |  ops/s |
|                                        50th percentile latency |   knn-search-10-50-acceptedAnswerId |  91.0864      |     ms |
|                                        90th percentile latency |   knn-search-10-50-acceptedAnswerId |  95.1255      |     ms |
|                                        99th percentile latency |   knn-search-10-50-acceptedAnswerId | 102.892       |     ms |
|                                       100th percentile latency |   knn-search-10-50-acceptedAnswerId | 105.84        |     ms |
|                                   50th percentile service time |   knn-search-10-50-acceptedAnswerId |  91.0864      |     ms |
|                                   90th percentile service time |   knn-search-10-50-acceptedAnswerId |  95.1255      |     ms |
|                                   99th percentile service time |   knn-search-10-50-acceptedAnswerId | 102.892       |     ms |
|                                  100th percentile service time |   knn-search-10-50-acceptedAnswerId | 105.84        |     ms |
|                                                     error rate |   knn-search-10-50-acceptedAnswerId |   0           |      % |
|                                                 Min Throughput | script-score-query-acceptedAnswerId |   1.94        |  ops/s |
|                                                Mean Throughput | script-score-query-acceptedAnswerId |   1.95        |  ops/s |
|                                              Median Throughput | script-score-query-acceptedAnswerId |   1.95        |  ops/s |
|                                                 Max Throughput | script-score-query-acceptedAnswerId |   1.95        |  ops/s |
|                                        50th percentile latency | script-score-query-acceptedAnswerId | 506.142       |     ms |
|                                        90th percentile latency | script-score-query-acceptedAnswerId | 521.093       |     ms |
|                                        99th percentile latency | script-score-query-acceptedAnswerId | 544.209       |     ms |
|                                       100th percentile latency | script-score-query-acceptedAnswerId | 552.22        |     ms |
|                                   50th percentile service time | script-score-query-acceptedAnswerId | 506.142       |     ms |
|                                   90th percentile service time | script-score-query-acceptedAnswerId | 521.093       |     ms |
|                                   99th percentile service time | script-score-query-acceptedAnswerId | 544.209       |     ms |
|                                  100th percentile service time | script-score-query-acceptedAnswerId | 552.22        |     ms |
|                                                     error rate | script-score-query-acceptedAnswerId |   0           |      % |
|                                                 Min Throughput |               knn-search-10-50-java |  10.48        |  ops/s |
|                                                Mean Throughput |               knn-search-10-50-java |  10.61        |  ops/s |
|                                              Median Throughput |               knn-search-10-50-java |  10.61        |  ops/s |
|                                                 Max Throughput |               knn-search-10-50-java |  10.72        |  ops/s |
|                                        50th percentile latency |               knn-search-10-50-java |  89.0557      |     ms |
|                                        90th percentile latency |               knn-search-10-50-java |  94.2473      |     ms |
|                                        99th percentile latency |               knn-search-10-50-java | 188.35        |     ms |
|                                       100th percentile latency |               knn-search-10-50-java | 227.795       |     ms |
|                                   50th percentile service time |               knn-search-10-50-java |  89.0557      |     ms |
|                                   90th percentile service time |               knn-search-10-50-java |  94.2473      |     ms |
|                                   99th percentile service time |               knn-search-10-50-java | 188.35        |     ms |
|                                  100th percentile service time |               knn-search-10-50-java | 227.795       |     ms |
|                                                     error rate |               knn-search-10-50-java |   0           |      % |
|                                                 Min Throughput |             script-score-query-java |   7.72        |  ops/s |
|                                                Mean Throughput |             script-score-query-java |   7.76        |  ops/s |
|                                              Median Throughput |             script-score-query-java |   7.76        |  ops/s |
|                                                 Max Throughput |             script-score-query-java |   7.78        |  ops/s |
|                                        50th percentile latency |             script-score-query-java | 125.497       |     ms |
|                                        90th percentile latency |             script-score-query-java | 132.617       |     ms |
|                                        99th percentile latency |             script-score-query-java | 138.362       |     ms |
|                                       100th percentile latency |             script-score-query-java | 139.082       |     ms |
|                                   50th percentile service time |             script-score-query-java | 125.497       |     ms |
|                                   90th percentile service time |             script-score-query-java | 132.617       |     ms |
|                                   99th percentile service time |             script-score-query-java | 138.362       |     ms |
|                                  100th percentile service time |             script-score-query-java | 139.082       |     ms |
|                                                     error rate |             script-score-query-java |   0           |      % |
|                                                 Min Throughput |                knn-search-10-50-css |  10.66        |  ops/s |
|                                                Mean Throughput |                knn-search-10-50-css |  10.8         |  ops/s |
|                                              Median Throughput |                knn-search-10-50-css |  10.82        |  ops/s |
|                                                 Max Throughput |                knn-search-10-50-css |  10.89        |  ops/s |
|                                        50th percentile latency |                knn-search-10-50-css |  88.3835      |     ms |
|                                        90th percentile latency |                knn-search-10-50-css |  90.6141      |     ms |
|                                        99th percentile latency |                knn-search-10-50-css |  95.0067      |     ms |
|                                       100th percentile latency |                knn-search-10-50-css | 100.368       |     ms |
|                                   50th percentile service time |                knn-search-10-50-css |  88.3835      |     ms |
|                                   90th percentile service time |                knn-search-10-50-css |  90.6141      |     ms |
|                                   99th percentile service time |                knn-search-10-50-css |  95.0067      |     ms |
|                                  100th percentile service time |                knn-search-10-50-css | 100.368       |     ms |
|                                                     error rate |                knn-search-10-50-css |   0           |      % |
|                                                 Min Throughput |              script-score-query-css |   5.29        |  ops/s |
|                                                Mean Throughput |              script-score-query-css |   6.23        |  ops/s |
|                                              Median Throughput |              script-score-query-css |   6.29        |  ops/s |
|                                                 Max Throughput |              script-score-query-css |   7           |  ops/s |
|                                        50th percentile latency |              script-score-query-css |  87.5079      |     ms |
|                                        90th percentile latency |              script-score-query-css |  90.4777      |     ms |
|                                        99th percentile latency |              script-score-query-css |  93.7029      |     ms |
|                                       100th percentile latency |              script-score-query-css |  94.0575      |     ms |
|                                   50th percentile service time |              script-score-query-css |  87.5079      |     ms |
|                                   90th percentile service time |              script-score-query-css |  90.4777      |     ms |
|                                   99th percentile service time |              script-score-query-css |  93.7029      |     ms |
|                                  100th percentile service time |              script-score-query-css |  94.0575      |     ms |
|                                                     error rate |              script-score-query-css |   0           |      % |
|                                                 Min Throughput |        knn-search-10-50-concurrency |  13.14        |  ops/s |
|                                                Mean Throughput |        knn-search-10-50-concurrency |  13.36        |  ops/s |
|                                              Median Throughput |        knn-search-10-50-concurrency |  13.38        |  ops/s |
|                                                 Max Throughput |        knn-search-10-50-concurrency |  13.52        |  ops/s |
|                                        50th percentile latency |        knn-search-10-50-concurrency |  70.4345      |     ms |
|                                        90th percentile latency |        knn-search-10-50-concurrency |  71.4835      |     ms |
|                                        99th percentile latency |        knn-search-10-50-concurrency |  72.8865      |     ms |
|                                       100th percentile latency |        knn-search-10-50-concurrency |  73.7157      |     ms |
|                                   50th percentile service time |        knn-search-10-50-concurrency |  70.4345      |     ms |
|                                   90th percentile service time |        knn-search-10-50-concurrency |  71.4835      |     ms |
|                                   99th percentile service time |        knn-search-10-50-concurrency |  72.8865      |     ms |
|                                  100th percentile service time |        knn-search-10-50-concurrency |  73.7157      |     ms |
|                                                     error rate |        knn-search-10-50-concurrency |   0           |      % |
|                                                 Min Throughput |      script-score-query-concurrency |  13.01        |  ops/s |
|                                                Mean Throughput |      script-score-query-concurrency |  13.25        |  ops/s |
|                                              Median Throughput |      script-score-query-concurrency |  13.28        |  ops/s |
|                                                 Max Throughput |      script-score-query-concurrency |  13.44        |  ops/s |
|                                        50th percentile latency |      script-score-query-concurrency |  70.6921      |     ms |
|                                        90th percentile latency |      script-score-query-concurrency |  71.8873      |     ms |
|                                        99th percentile latency |      script-score-query-concurrency |  73.6924      |     ms |
|                                       100th percentile latency |      script-score-query-concurrency |  74.5011      |     ms |
|                                   50th percentile service time |      script-score-query-concurrency |  70.6921      |     ms |
|                                   90th percentile service time |      script-score-query-concurrency |  71.8873      |     ms |
|                                   99th percentile service time |      script-score-query-concurrency |  73.6924      |     ms |
|                                  100th percentile service time |      script-score-query-concurrency |  74.5011      |     ms |
|                                                     error rate |      script-score-query-concurrency |   0           |      % |

```

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/CartesianShapeValues.java [hunk 11]
```diff
@@ -67,6 +67,7 @@ public CartesianShapeValue() {
             super(CoordinateEncoder.CARTESIAN, CartesianPoint::new);
         }
 
+        @SuppressWarnings("this-escape")
```
Comment (jimczi) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/CartesianShapeValues.java:70]: looks unrelated?

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/GeoShapeValues.java [hunk 12]
```diff
@@ -70,6 +70,7 @@ public GeoShapeValue() {
             this.tile2DVisitor = new Tile2DVisitor();
         }
 
+        @SuppressWarnings("this-escape")
```
Comment (jimczi) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/GeoShapeValues.java:73]: same here?

File: docs/reference/search/search-your-data/knn-search.asciidoc [hunk 13]
```diff
@@ -242,6 +242,117 @@ POST byte-image-index/_search
...
+{
+  "knn": {
+    "field": "image-vector",
+    "query_vector": [0.1, -2],
+    "k": 15,
+    "num_candidates": 100
+  },
+  "fields": [ "title" ],
+  "rescore": {
+    "window_size": 10,
```
Comment (jimczi) [inline docs/reference/search/search-your-data/knn-search.asciidoc:333]: I think the `window_size` should be set to 15 to match the intent with the `k` value above?

File: server/src/main/java/org/elasticsearch/action/ingest/SimulateIndexResponse.java [hunk 14]
```diff
@@ -31,13 +31,15 @@ public class SimulateIndexResponse extends IndexResponse {
     private final BytesReference source;
     private final XContentType sourceXContentType;
 
+    @SuppressWarnings("this-escape")
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/action/ingest/SimulateIndexResponse.java:34]: Unrelated?

**Review by jimczi — COMMENTED** [review]
LGTM!

File: docs/reference/search/search-your-data/knn-search.asciidoc (hunk 13 above)
Comment (benwtrent) [reply docs/reference/search/search-your-data/knn-search.asciidoc:333]: It honestly doesn't matter. I would think you would get a larger k, and then rerank some sub-set of those.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/GeoShapeValues.java (hunk 12 above)
Comment (benwtrent) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/GeoShapeValues.java:73]: Again, this is something that I fixed in main but is breaking my local build. For sanity, I fixed it here. I will be a no-op once main is merged here

File: server/src/main/java/org/elasticsearch/action/ingest/SimulateIndexResponse.java (hunk 14 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/action/ingest/SimulateIndexResponse.java:34]: It will go away when we merge main, but it is breaking my local testing.

File: server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java (hunk 4 above)
Comment (kderusso) [reply server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:180]: Fair.

**Review by kderusso — APPROVED** [review]
Very nice work!

**Comment by benwtrent** [comment]
> is there anything that is in the way of making this the default?

What would the API look like to disable it? If the user provides any HNSW params and doesn't specify quantization, does this mean we should default?

**Comment by jpountz** [comment]
> What would the API look like to disable it? If the user provides any HNSW params and doesn't specify quantization, does this mean we should default?

Ideally defaults would reflect the knn search tuning guide, so my current thinking right now is to have something like:
 - quantization enabled automatically when `dim >= 384`
 - have a `quantization_options.type = "none"` or something along these lines that can be used to disable quantization explicitly
 - indeed quantization would remain enabled if a user overrides HNSW options without touching `quantization_options.type `

File: docs/reference/mapping/types/dense-vector.asciidoc [hunk 15]
```diff
@@ -212,14 +246,36 @@ properties must be defined.
 The type of kNN algorithm to use. Currently only `hnsw` is supported.
 
 `m`:::
-(Required, integer)
+(Optional, integer)
```
Comment (ChrisHegarty) [inline docs/reference/mapping/types/dense-vector.asciidoc:249]: The previous paragraph has a sentence that needs to be updated to reflect the new optionality of these parameters. The sentence is:

`When index_options is provided, all of its properties must be defined.`

File: docs/reference/mapping/types/dense-vector.asciidoc [hunk 16]
```diff
@@ -111,6 +111,40 @@ PUT my-index-2
 efficient kNN search. Like most kNN algorithms, HNSW is an approximate method
 that sacrifices result accuracy for improved speed.
 
+[[dense-vector-quantization]]
+==== Automatically quantize vectors for kNN search
+
+The `dense_vector` type supports quantization to reduce the memory footprint required when searching `float` vectors.
```
Comment (ChrisHegarty) [inline docs/reference/mapping/types/dense-vector.asciidoc:117]: "searching"  - this is specifically approx kNN, right? Not exact brute force search. If so, then maybe clarify this or just link to `knn-search.html#approximate-knn` ?

File: docs/reference/mapping/types/dense-vector.asciidoc [hunk 17]
```diff
@@ -212,14 +246,36 @@ properties must be defined.
...
-(Required, integer)
+(Optional, integer)
 The number of candidates to track while assembling the list of nearest
 neighbors for each new node. Defaults to `100`.
+
+`quantization_options`:::
+(Optional, object)
+An optional section that configures the quantization configuration. The
+quantization configuration is used to reduce the memory footprint of the
+index. Only `byte` quantization is currently supported and can only be configured if
```
Comment (ChrisHegarty) [inline docs/reference/mapping/types/dense-vector.asciidoc:262]: I get confused by this; " memory footprint of the index".  Quantization retains the original floats, and also has the quantized byte values, so the index is 1.25 times the size, no?   It's really just a "approx knn searches only result in loading the quantized vectors", right ?

File: docs/reference/search/search-your-data/knn-search.asciidoc [hunk 18]
```diff
@@ -242,6 +242,117 @@ POST byte-image-index/_search
...
+  },
+  "fields": [ "title" ]
+}
+----
+// TEST[continued]
+// TEST[s/"k": 10/"k": 3/]
+// TEST[s/"num_candidates": 100/"num_candidates": 3/]
+
+Since the original `float` vectors are still retained in the index, you can optionally use them for re-scoring. This will
+do the heavy query against the indexed vectors, and then you can get the absolute nearest neighbors by re-scoring.
```
Comment (ChrisHegarty) [inline docs/reference/search/search-your-data/knn-search.asciidoc:319]: "heavy query" implies no memory footprint improvements, right? The original float vectors are loaded.

**Review by ChrisHegarty — COMMENTED** [review]
This is a great improvement.

I left some nit picky comments on the docs; mostly do to with my own confusion - this is not an easy concept to grasp, and I believe a tightening of the wording will help folks with it. 

What I struggle with is the use of `byte` while there is already a `byte` element_type. How to we want users to think about this: quantization is just an implementation detail when using float element_type that improves runtime memory footprint OR quantization is somehow coercing floats to bytes, and the user now need to think of bytes.  I would think the former, which kinda relates to other comments in this thread; maybe refer to the quantization as `int8` or `1-byte integer` value ? Now users of float do not need to think of the `byte` element_type.

File: docs/reference/mapping/types/dense-vector.asciidoc (hunk 17 above)
Comment (benwtrent) [reply docs/reference/mapping/types/dense-vector.asciidoc:262]: Memory footprint for approximate search is lower. The disk footprint is now higher. I can adjust the docs to be clearer.

File: docs/reference/search/search-your-data/knn-search.asciidoc (hunk 18 above)
Comment (benwtrent) [reply docs/reference/search/search-your-data/knn-search.asciidoc:319]: Docs here are unclear obviously. I mean to say that the expensive query is done via approximate search against smaller foot print.

**Comment by benwtrent** [comment]
@jpountz @jimczi 

>  is there anything that is in the way of making this the default?

I would not be comfortable making it the default right away. I really want more testing before we do this. But, we should indeed design the API in a way that makes this possible.


For @jpountz 's option,
```
{
  "mappings": {
    "properties": {
      "my_vector": {
        "type": "dense_vector",
        "index": true,
        "index_options": {
          "type": "hnsw",
          "quantization": true, // like "index"
          "quantization_options": { // like "index_options"
            "type": "byte",
            "confidence_interval": 0.9
          }
        }
      }
    }
  }
}
```

For @jimczi option, where we add a new index type, that could also work.

For setting confidence interval, it would be
```
{
  "mappings": {
    "properties": {
      "my_vector": {
        "type": "dense_vector",
        "index": true,
        "index_options": {
          "type": "int8_hnsw",
          "confidence_interval": 0.9
        }
      }
    }
  }
}
```
For things that don't support that parameter, we would throw.

What do you think @jpountz ^ For things in the future `pq` `int4`, it would be `pq_hnsw` or `int4_hnsw` and for flat it would be `int8_flat`, etc.

@ChrisHegarty 's comment about `byte` vs `int8`.

Yes, I should update the quantization value name from `byte` to `int8` to prevent term overloading and confusion.

**Comment by jpountz** [comment]
> What do you think @jpountz ^ For things in the future pq int4, it would be pq_hnsw or int4_hnsw and for flat it would be int8_flat, etc.

The question that this raises to me is whether "flat" should be considered as a special index type, or as a lack of index. I had initially assumed that flat storage would be enabled by setting `index: false` because it seems more consistent with other fields we support as flat storage feels like doc values, but on the other hand other vector search libraries do seem to consider flat storage as a form of index. If we want to only enable flat storage through `index: false` then it would be awkward to configure quantization as part of index options? And if `flat` is a special form of indexing, then what does `index: false` do?

**Comment by benwtrent** [comment]
We need to support quantization over various indexing methodologies, including "flat". 

I think using "index: false" as a flat index mistake. It should only be useful for scripting.

It is really weird to have "index: false" and then configure what you are "not" indexing :facepalm:. 

I think we need a new "flat" index kind that requires the similarity to be configured, and allows quantization. Admittedly, just "flat" isn't much better than "index: false", but it will clean up the query API really nicely (as we know the similarity that's configured).

Another option is to have quantization live as a top level thing. But all this configuration is getting unwieldy.

I am starting to like @jimczi's suggestion more and more.

File: docs/reference/mapping/types/dense-vector.asciidoc (hunk 17 above)
Comment (mayya-sharipova) [inline docs/reference/mapping/types/dense-vector.asciidoc:262]: Do you think we should add a sentence about disk usage increase with `quantization_options`?

**Comment by benwtrent** [comment]
@mayya-sharipova what do you think of @jimczi's suggested API? Adding a new "int8_hnsw" index type.

File: docs/reference/mapping/types/dense-vector.asciidoc [hunk 19]
```diff
@@ -212,14 +246,36 @@ properties must be defined.
...
+`element_type` is `float` and `index: true`.
+.Properties of `quantization_options`
+[%collapsible%open]
+=====
+
+`type`:::
+(Required, string)
+The type of quantization. Currently only `byte` is supported.
+
+`confidence_interval`:::
```
Comment (mayya-sharipova) [inline docs/reference/mapping/types/dense-vector.asciidoc:272]: Is `confidence_interval` applicable for other type of quantization that we may introduce (e.g. product quantization)?

Comment (benwtrent) [reply docs/reference/mapping/types/dense-vector.asciidoc:272]: @mayya-sharipova its only applicable to things that are scalar quantization. So, probably only to `int4` and `int8`

**Comment by jpountz** [comment]
So if I read your suggestion correctly, we'd treat `flat` as a form of index and parse `index: false` as meaning "flat index without quantization" for compatibility? That would work for me. I like that the suggested API from @jimczi is less verbose, much nicer to read.

**Comment by mayya-sharipova** [comment]
+1 on  @jimczi's proposal.  

+1 also to add flat as an index_option: (may be in another PR):

```js
"similarity": "l2_norm",
"index_options": {
  "type": "flat"
}
```

**Review by jimczi — APPROVED** [review]
LGTM, thanks for the iteration

**Review by mayya-sharipova — APPROVED** [review]
Thanks @benwtrent, great work.

New changes to `index_options` look good as well.

**Review by ChrisHegarty — APPROVED** [review]
LGTM

File: docs/reference/mapping/types/dense-vector.asciidoc (hunk 17 above)
Comment (Rassyan) [reply docs/reference/mapping/types/dense-vector.asciidoc]: @benwtrent 
In the context of the current implementation of quantization in Elasticsearch, if we were to store only the quantized byte vectors and not retain the original float32 vectors, would it be correct to assume that both disk and memory footprints could be reduced to approximately 25% of the original size? This approach seems like it could offer significant savings. Could you elaborate on why this option is not currently supported? Are there plans to introduce such a feature in future releases? Thank you for your insights.
